autorest/adal/token.go

```diff
@@ -686,8 +693,8 @@ func NewServicePrincipalTokenFromAuthorizationCode(oauthConfig OAuthConfig, clie
 	)
 }
 
-// NewServicePrincipalTokenFromFederatedToken creates a ServicePrincipalToken from the supplied federated OIDC JWT.
-func NewServicePrincipalTokenFromFederatedToken(oauthConfig OAuthConfig, clientID string, jwt string, resource string, callbacks ...TokenRefreshCallback) (*ServicePrincipalToken, error) {
+// NewServicePrincipalTokenFromFederatedToken creates a ServicePrincipalToken from the supplied federated OIDC JWTCallback.
+func NewServicePrincipalTokenFromFederatedToken(oauthConfig OAuthConfig, clientID string, jwtCallback JWTCallback, resource string, callbacks ...TokenRefreshCallback) (*ServicePrincipalToken, error) {
```

Comment:
This is a breaking change.

Reply:
I doubt whether there is any user would call this `NewServicePrincipalTokenFromFederatedToken` since the token would expire after 24 hours, that's more like a bug

Reply:
This bug actually only happens in such scenario:

For applications running on AKS, they need to use the JWT, which is generated by apiserver and mount to the Pod by kubelet , to exchange OAuth token from Azure identity platform when there is any request to arm **at the first time**. Azure will not only return OAuth token for authenticating, it will also return a `refresh_token` for the application to refresh Oauth token. ADAL will use the JWT to request Oauth token and `refresh_token` when there is a request to arm at the first time, and then it will use `refresh_token` to refresh the OAuth token if it is expired and the JWT should be useless.

However, application may have no request to arm for a very long time after it is up (or after master-slave switch). If the JWT is expired, apiserver will refresh the JWT and write to file mounted in the Pod. Under current implementation, the jwt we used to exchange token from Azure identity platform cannot be refreshed since we cannot read the JWT file from the file dynamically. So after the JWT is expired, request to exchange token would be rejected.